src/stepfunctions/steps/sagemaker.py

```diff
@@ -103,6 +110,10 @@ def __init__(self, state_id, estimator, job_name, data=None, hyperparameters=Non
         if isinstance(job_name, (ExecutionInput, StepInput)):
             parameters['TrainingJobName'] = job_name
 
+        if output_path is not None:
+            if isinstance(output_path, (ExecutionInput, StepInput)):
+                parameters['OutputDataConfig']['S3OutputPath'] = output_path
```

Comment:
Looking at the [CreateTrainingJob API docs](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html) there's a lot of the API surface we aren't covering. e.g. `InputDataConfig`, `ResourceConfig`, etc. I am also see a few other parameters that accept an S3OutputPath, like [DebugRuleConfigurations](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html#sagemaker-CreateTrainingJob-request-DebugRuleConfigurations).

I think there are two approaches we can take here which applies to all Tasks:
1. Selectively expose properties can use placeholders
2. Change how Task constructs the ASL to always replace placeholders in parameters (and result_selector when we add it)

Reply:
Actually, it looks like that already happens in the JSON serialization:
https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/main/src/stepfunctions/steps/states.py#L51-L77

So the issue is more about making all arguments we expose from the TrainingStep compatible with placeholders.

Reply:
Yes, agreed!
We would need to pass all TrainingStep potential placeholder arguments to parameters so that they can be replaced during JSON serialization.

Reply:
> So the issue is more about making all arguments we expose from the TrainingStep compatible with placeholders.

yes, that's what I had in mind as well for implementation here (and related issues)

Reply:
Let's focus on data and output_path for this PR and make the other arguments compatible with placeholders in another. There is a [feature request ](https://github.com/aws/aws-step-functions-data-science-sdk-python/issues/139) to make ProcessingStep arguments compatible with placeholders as well - we could do this on a higher level (for ex [Task](https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/main/src/stepfunctions/steps/states.py#L583)) so that there is less repetitiveness.

Reply:
So my comment now is actually about what `output_path` refers to. As I mentioned earlier, CreateTrainingJob has 6 instances of S3OutputPath in different parameters. Is `output_path` descriptive enough or should the name be more specific? I believe the current name was just chosen from https://github.com/aws/aws-step-functions-data-science-sdk-python/issues/98

Reply:
The defined parameters are passed down to the [sagemaker.estimator.EstimatorBase](https://sagemaker.readthedocs.io/en/stable/api/training/estimators.html?highlight=EstimatorBase#sagemaker.estimator.EstimatorBase). `output_path` name was used here to concord with the `output_path` in [EstimatorBase](https://sagemaker.readthedocs.io/en/stable/api/training/estimators.html?highlight=EstimatorBase#sagemaker.estimator.EstimatorBase). I feel it might add confusion to name it something else, but I am not strongly opinionated about it - we could use a more descriptive name and mention in the docs that it maps to [EstimatorBase](https://sagemaker.readthedocs.io/en/stable/api/training/estimators.html?highlight=EstimatorBase#sagemaker.estimator.EstimatorBase)'s `output_path`.

Would `result_path` be a better option?

Reply:
Looking at the [CreateTrainingJob](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateTrainingJob.html) docs you mentioned earlier, would `output_data_config_path` work? Since `OutputDataConfig` is the property we update in the parameters, this could be a better option.

Reply:
`output_data_config_path` seems alright to me

Both `result_path` and `output_path` conflict with existing Task properties of the same name, so we can't use those. Sagemaker steps extend `Task` and the caller should still be able to specify those fields like any other Task.
https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/6091932448572244801362419034414b48c9f4e6/src/stepfunctions/steps/states.py#L601-L602

Reply:
Agreed - `result_path` and `output_path` should not be used if they are conflicting with Task properties. 
`output_path` will be renamed to `output_data_config_path` in the next commit.